pkg/ccl/sqlproxyccl/server.go

```diff
@@ -349,6 +395,52 @@ func (s *Server) AwaitNoConnections(ctx context.Context) <-chan struct{} {
 	return c
 }
 
+// shouldLogError returns true if an error should be logged, or false otherwise.
+// The goal is to only throttle high-frequency errors for every (IP, tenantID)
+// pair.
+func (s *Server) shouldLogError(err error, conn net.Conn, reqTags map[string]interface{}) bool {
+	// Always log non high-frequency errors.
+	if !errors.Is(err, highFreqErrorMarker) {
+		return true
+	}
+
+	// Request hasn't been populated with a tenant ID yet, so log them.
+	tenID, hasTenantID := reqTags["tenant"]
+	if !hasTenantID {
+		return true
+	}
+
+	// Tenant ID must be a string, or else there has to be a bug in the code.
+	// Instead of panicking, we'll skip throttling.
+	tenantIDStr, ok := tenID.(string)
+	if !ok {
```

Comment:
should we meta-log this "this should never happen (TM)" case somehow?

Reply:
Done.